qpdldecode.1
'\" t
'\"
'\"	DO NOT EDIT! This file is generated from qpdldecode.1in
'\"
.TH qpdldecode 1 "Sun Apr 25 16:22:30 2021" "qpdldecode 0.0"
'\"
'\"==========================================================================
'\"	STRINGS and MACROS
'\"==========================================================================
'\"
'\"	Define strings for special characters that nroff doesn't have
'\"
'\"	N.B. using .if cua to test for special characters did not work.
'\"
.if !'\*[.T]'ps' .ds ua up
.if  '\*[.T]'ps' .ds ua \(ua
.if !'\*[.T]'ps' .ds da down
.if  '\*[.T]'ps' .ds da \(da
.if !'\*[.T]'ps' .ds <- left
.if  '\*[.T]'ps' .ds <- \(<-
.if !'\*[.T]'ps' .ds -> right
.if  '\*[.T]'ps' .ds -> \(->
'\"
'\"     bx - box a Courier string for making keycaps
'\"
'\"	N.B. this mess is to make the line drawing come out only
'\"	if we are really generating postscript
'\"
.de bx
.ie !'\*[.T]'ps' \{\
.	RB [ \\$1 ]\\$2
.\}
.el \{\
.	ie !r ps4html \{\
'\" \(br\|\s-1\f(CB\\$1\fP\s+1\|\(br\l'|0\(rn'\l'|0\(ul'
.		ft CW
.		nr par*bxw \w'\\$1'+.4m 
\Z'\v'.25m'\D'l 0 -1m'\D'l \\n[par*bxw]u 0'\D'l 0 1m'\D'l -\\n[par*bxw]u 0''\
\Z'\h'.2m'\s-1\\$1\s+1'\
\h'\\n[par*bxw]u'\\$2
.		ft P
.	\}
.	el \{\
.		RB [ \\$1 ]\\$2
.	\}
.\}
..
'\"
'\" strings to set current color (null with old groff)
'\"
.if mred .ds red \m[red]
.if mgreen .ds green \m[green]
.if mblue .ds blue \m[blue]
.if mblack .ds black \m[black]
.if mblack .ds mP \mP
'\"
'\" fix for grotty + xterm. We call for orange, grotty outputs yellow,
'\" but xterm displays yellow as orange.  The cycle is complete.
'\"
.if n .defcolor orange rgb #ffff00
'\"
'\" color <color> - set the current color (ignores request with old groff)
'\"
.de color
.if mred \m[\\$1]\c
..
'\"
'\" colorword <color> <word> - colorize a word (ignored by old groff)
'\"
.de colorword
.ie m\\$1 \m[\\$1]\\$2\mP\c
.el \\$2\c
..
'\"
'\" colbox <fg> <bg> <word> - colorize a word in a filled box
'\"
.de colbox
.ie mred \M[\\$2]\
\v'+.167v'\
\D'P 0 -0.9v  \w'\\$3'u 0  0 +0.9v   -\w'\\$3'u 0'\
\v'-.167v'\
\m[\\$1]\\$3\mP\MP
.el \\$3\c
..
'\"
'\"	Macros for doing pdfmarks
'\"
.de specialps
.if  '\*[.T]'ps' \\k_\X'ps: \\$*'\h'|\\n_u'\c
..
'\"
'\" pdfmark PDFMARKCODE
'\"
.ds pdfmarks
.if d pdfmarks \{\
.de pdfmark
.	specialps exec [\\$1 pdfmark
..
'\"
'\" pdfdest LINKNAME
'\"
.de pdfdest
.pdfmark "/Dest /\\$1 /View [/XYZ -5 PL null] /DEST"
..
'\"
'\" pdfbookmark COUNT LINKNAME STRING
'\"
.de pdfbookmark
.   pdfmark "/View [/XYZ 44 730 1.0] /Count \\$1 /Dest /\\$2 /Title (\\$3) /OUT"
..
'\"
'\"	Define the SH and SS macros to save pdfmark information
'\"	in "arrays" of numbers and strings.
'\"
.if !r rr_n \{\
.nr rr_n 0 1
.am SH
.	nr rr_levels!\\n+[rr_n] 2
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.am SS
.	nr rr_levels!\\n+[rr_n] 3
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.\}
'\"
'\"	Called at the end of the document to generate the pdfmark outline
'\"
.de pdf_outline
.nr rr_levels!\\n+[rr_n] 1
.nr rr_i 0 1
.while \\n+[rr_i]<\\n[rr_n] \{\
.   nr rr_ip1 \\n[rr_i]+1
.   nr rr_count 0
.   if \\n[rr_levels!\\n[rr_ip1]]>\\n[rr_levels!\\n[rr_i]] \{\
.       nr rr_j \\n[rr_i] 1
.       while \\n+[rr_j]<\\n[rr_n] \{\
.           if \\n[rr_levels!\\n[rr_j]]<=\\n[rr_levels!\\n[rr_i]] \{\
.               break
.           \}
.           if \\n[rr_levels!\\n[rr_j]]==(\\n[rr_levels!\\n[rr_i]]+1) \{\
.               nr rr_count \\n[rr_count]+1
.           \}
.       \}
.   \}
.   ds hhh \\*[rr_labels!\\n[rr_i]]
.   pdfbookmark -\\n[rr_count] Link\\n[rr_i] "\\*[hhh]"
.\}
..
'\"
'\" Some postscript to make pdfmarks harmless on old interpreters...
'\"
.specialps "def /pdfmark where {pop} {userdict /pdfmark /cleartomark load put} ifelse"
'\"
'\" Force display of Bookmarks in Acrobat when document is viewed.
'\"
.pdfmark "[/PageMode /UseOutlines /Page 1 /View [/XYZ null null null] /DOCVIEW"
'\"
'\" Output the document info in pdfmarks
'\"
.pdfmark "\
	/Title (\*[an-title](\*[an-section])) \
	/Subject (\*[an-title] Manual Page) \
	/Author (Rick Richardson) \
	/Keywords (printing) \
	/Creator (groff \n(.x.\n(.y.\n(.Y -man) \
	/CreationDate (\*[an-extra1]) \
	/ModDate (\*[an-extra1]) \
	/DOCINFO"
\}
'\"
'\" The manual page name is only 1st level mark
'\"
.nr rr_levels!\n+[rr_n] 1
.ds rr_labels!\n[rr_n] \*[an-title](\*[an-section])
.pdfdest Link\n[rr_n]
'\"
'\"==========================================================================
'\"	MANUAL PAGE SOURCE
'\"==========================================================================
.SH NAME
qpdldecode \- Decode a QPDL stream into human readable form.
.SH SYNOPSIS
.B qpdldecode
.RI [ options "] <" qpdl-file
.SH DESCRIPTION
.B qpdldecode
decodes a QPDL stream into human readable form.
Only the JBIG compression format (0x13) is handled.
.P
An QPDL stream is the printer language used by the
Samsung CLP-300, CLP-600, CLX-3160 and the
Xerox Phaser 6110
printers.

.SH COMMAND LINE OPTIONS
These are the options that can appear on the command line.
.TP
.BI \-d\0 basename
Basename of .pbm file for saving decompressed planes.
.TP
.BI \-h
Print hex file offsets.
.TP
.BI \-o
Print file offsets.
.TP
.BI \-D\0 level
Set Debug level [0].

.SH EXAMPLES
Decode an QPDL stream file created by foo2qpdl.

.nf
.ft CW
     0:    \\033%-12345X@PJL DEFAULT SERVICEDATE=20070212
    2c:    @PJL SET USERNAME="Unknown"
    49:    @PJL SET JOBNAME="testpage.pdf"
    6a:    @PJL SET COLORMODE=COLOR
    84:    @PJL SET PAPERTYPE = NORMAL
    a1:    @PJL ENTER LANGUAGE = QPDL
    bd:    RECTYPE 0x0  len=17
                res=600, copies=1, papersize=letter(0), w=2550, h=3300
                papersource=auto, unk=0, duplex=0:0, unk=0,2,  unk=268(0x10c)
    ce:    RECTYPE 0xc  len=68(0x44)
                stripe=0, WB=1248(0x4e0), H=128(0x80), plane=4, comp=0x13,
                len=56(0x38)
                magic=0x39abcdef, len=20(0x14), unk=0,0,0,0,0,0,
                checksum=0x356
                DL = 0, D = 0, P = 1, - = 0, XY = 9984 x 6400
                L0 = 6400, MX = 0, MY = 0
                Order   = 0
                Options = 72  LRLTWO TPBON
                1 stripes, 0 layers, 1 planes
   112:    RECTYPE 0xc  len=68(0x44)
                stripe=0, WB=1248(0x4e0), H=128(0x80), plane=1, comp=0x13,
                len=56(0x38)
                magic=0x39abcdef, len=20(0x14), unk=0,0,0,0,0,0,
                checksum=0x356
                DL = 0, D = 0, P = 1, - = 0, XY = 9984 x 6400
                L0 = 6400, MX = 0, MY = 0
                Order   = 0
                Options = 72  LRLTWO TPBON
                1 stripes, 0 layers, 1 planes
   156:    RECTYPE 0xc  len=68(0x44)
                stripe=0, WB=1248(0x4e0), H=128(0x80), plane=2, comp=0x13,
                len=56(0x38)
                magic=0x39abcdef, len=20(0x14), unk=0,0,0,0,0,0,
                checksum=0x356
                DL = 0, D = 0, P = 1, - = 0, XY = 9984 x 6400
                L0 = 6400, MX = 0, MY = 0
                Order   = 0
                Options = 72  LRLTWO TPBON
                1 stripes, 0 layers, 1 planes
   19a:    RECTYPE 0xc  len=68(0x44)
                stripe=0, WB=1248(0x4e0), H=128(0x80), plane=3, comp=0x13,
                len=56(0x38)
                magic=0x39abcdef, len=20(0x14), unk=0,0,0,0,0,0,
                checksum=0x356
                DL = 0, D = 0, P = 1, - = 0, XY = 9984 x 6400
                L0 = 6400, MX = 0, MY = 0
                Order   = 0
                Options = 72  LRLTWO TPBON
                1 stripes, 0 layers, 1 planes
   1de:    RECTYPE 0xc  len=77488(0x12eb0)
                stripe=1, WB=1248(0x4e0), H=128(0x80), plane=1, comp=0x13,
                len=77476(0x12ea4)
                magic=0x39abcdef, len=77440(0x12e80), unk=2000000,0,0,0,0,0,
                checksum=0x9326d7
 1308e:    RECTYPE 0xc  len=77680(0x12f70)
                stripe=1, WB=1248(0x4e0), H=128(0x80), plane=2, comp=0x13,
                len=77668(0x12f64)
                magic=0x39abcdef, len=77632(0x12f40), unk=2000000,0,0,0,0,0,
                checksum=0x9367e5

ChangeLog
2020-06-10	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: Add PREFIX=/usr or PREFIX=/usr/local to cc rules

2020-05-05:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: SaveToner implemented for color pages,
	  thanks to bedi@3mtech.sk .

2020-04-26:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug with SaveToner

2020-02-07:	Rick Richardson <rick.richardson@comcast.net>
	* Add Cassette 2 to foo2oak-z1-InputSlot.xml (KM-1635)

2019-09-09:	Rick Richardson <rick.richardson@comcast.net>
	* Add SmartInstallDisable-Tool.run since it was gone from
	      http://hplipopensource.com/hplip-web/smartinstall/
	  See HP SMART INSTALL NOTES in the INSTALL doc.

2019-04-13:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: Add install-aa to disable AppArmor on SUSE

2018-09-10:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:	revert the 09-06 change.

2018-09-06:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:
	    Use the older foomatic-ppdfile which only has postscript,
	    not postscript and pdf, for making ppd's.

2017-12-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printers: HP LaserJet Pro M12a, M12w

2017-03-20	Rick Richardson <rick.richardson@comcast.net>
	* Add ddstdecode, for decoding some Ricoh printers.

2016-10-21	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: install-hotplug-prog
	    Add /lib/systemd/systemd-udevd to the list for Ubuntu 16.

2016-09-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3040 using the foo2hbpl2 driver.

2016-08-30	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3010 using the foo2hbpl2 driver.

2016-03-13	Rick Richardson <rick.richardson@comcast.net>
	* icc2ps/lcms.h: USE_CUSTOM_SWAB defined.

2015-10-24	Rick Richardson <rick.richardson@comcast.net>
	* foomatic-db/printer/Epson-AcuLaser_CX17NF.xml:
		Change margins on Epson AcuLaser CX17NF.

2015-10-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Epson AcuLaser CX17NF using the foo2hbpl2 driver.

2015-07-04	pavel.shvagirev	<pavel.shvagirev@gmail.com>
	* Modify: hplj10XX to not call us twice on Ubuntu 14.

2015-05-11	Rick Richardson <rick.richardson@comcast.net>
	* Bug in foomatic-db/printer/Xerox-Phaser_6115MFP.xml . Should use
	  foo2lava driver. Thanks to user "ophers".

2015-02-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Oki C511 using foo2hiperc.

2014-12-21	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: make it work with OS X 10.10

2014-12-04	Rick Richardson <rick.richardson@comcast.net>
	* make cups: add "Sandboxing Relaxed" to cups-files.conf iff cups >= 2

2014-09-25	Rick Richardson <rick.richardson@comcast.net>
	* foo2hblp2, foo2hblp2-wrapper, foomatic et al:
	    Add paper size A5.
	* foo2hblp2 foomatic db entries:
	    Changed resolution to 600x600, 1200x600, and 1200x1200.
	    Note: need to recreate the printers if you want more choices!

2014-05-19	Rick Richardson <rick.richardson@comcast.net>
	* HP CLJ CP1215:  Adjust margins: *HWMargins: 18 14.40 18 14

2014-05-04	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2-wrapper: Change PAPER=21 to PAPER=0 for custom page sizes.

2014-04-17	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Fuji Xerox DocuPrint CM215 MFP using the foo2hbpl2
	driver.

2014-04-08	Rick Richardson <rick.richardson@comcast.net>
	* Upgrade to JBIGKIT 2.1

2014-04-01	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2 foomatic-db files and PPDs:
	    Bug with default PageSize.
	    PLEASE delete and recreate the foo2hbpl2 printers.

2014-03-22	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: another redundancy from Dave Coffin

2014-03-18	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More envelope sizes.

2014-03-17	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug, was using A4(1) instead of letter(4) by default

qpdldecode.c
/*
 * $Id: qpdldecode.c,v 1.41 2014/01/24 19:25:47 rick Exp $
 */

/*b
 * Copyright (C) 2003-2006  Rick Richardson
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * Author: Rick Richardson <rick.richardson@comcast.net>
b*/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#include "jbig.h"

/*
 * Global option flags
 */
int	Debug = 0;
char	*RawFile;
char	*DecFile;
int	PrintOffset = 0;
int	PrintHexOffset = 0;

void
debug(int level, char *fmt, ...)
{
    va_list ap;

    if (Debug < level)
	return;

    va_start(ap, fmt);
    vfprintf(stderr, fmt, ap);
    va_end(ap);
}

int
error(int fatal, char *fmt, ...)
{
	va_list ap;

	fprintf(stderr, fatal ? "Error: " : "Warning: ");
	if (errno)
	    fprintf(stderr, "%s: ", strerror(errno));
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	if (fatal > 0)
	    exit(fatal);
	else
	{
	    errno = 0;
	    return (fatal);
	}
}

void
usage(void)
{
    fprintf(stderr,
"Usage:\n"
"	qpdldecode [options] < qpdl-file.prn\n"
"\n"
"	Decode a QPDL stream into human readable form.\n"
"\n"
"	A Quick Page Description Langauge (QPDL) is the printer language"
"	used by some Samsung printers, such as the CLP-600n.\n"
"\n"
"	Also known as SPLC."
"\n"
"Options:\n"
"       -d basename Basename of .pbm file for saving decompressed planes\n"
//"       -r basename Basename of .jbg file for saving raw planes\n"
"       -o          Print file offsets\n"
"       -h          Print hex file offsets\n"
"       -D lvl      Set Debug level [%d]\n"
    , Debug
    );

    exit(1);
}

static int
getBEdword(char buf[4])
{
    unsigned char *b = (unsigned char *) buf;
    return (b[0] << 24) | (b[1] << 16) | (b[2] << 8) | (b[3] << 0);
}

static int
getLEdword(char buf[4])
{
    unsigned char *b = (unsigned char *) buf;
    return (b[3] << 24) | (b[2] << 16) | (b[1] << 8) | (b[0] << 0);
}

static int
getBEword(char buf[2])
{
    unsigned char *b = (unsigned char *) buf;
    return (b[0] << 8) | (b[1] << 0);
}

void
print_bih(unsigned char bih[20])
{
    unsigned int xd, yd, l0;

    xd = (bih[4] << 24) | (bih[5] << 16) | (bih[6] << 8) | (bih[7] << 0);
    yd = (bih[8] << 24) | (bih[9] << 16) | (bih[10] << 8) | (bih[11] << 0);
    l0 = (bih[12] << 24) | (bih[13] << 16) | (bih[14] << 8) | (bih[15] << 0);

    printf("		DL = %d, D = %d, P = %d, - = %d, XY = %d x %d, "
	    "%s\n",
	    bih[0], bih[1], bih[2], bih[3], xd, yd,
	    (xd % 256) ? "*** xd%256 != 0!" : "");

    printf("		L0 = %d, MX = %d, MY = %d\n",
	l0, bih[16], bih[17]);

    printf("		Order   = %d %s%s%s%s%s\n", bih[18],
	bih[18] & JBG_HITOLO ? " HITOLO" : "",
	bih[18] & JBG_SEQ ? " SEQ" : "",
	bih[18] & JBG_ILEAVE ? " ILEAVE" : "",
	bih[18] & JBG_SMID ? " SMID" : "",
	bih[18] & 0xf0 ? " other" : "");

    printf("		Options = %d %s%s%s%s%s%s%s%s\n", bih[19],
	bih[19] & JBG_LRLTWO ? " LRLTWO" : "",
	bih[19] & JBG_VLENGTH ? " VLENGTH" : "",
	bih[19] & JBG_TPDON ? " TPDON" : "",
	bih[19] & JBG_TPBON ? " TPBON" : "",
	bih[19] & JBG_DPON ? " DPON" : "",
	bih[19] & JBG_DPPRIV ? " DPPRIV" : "",
	bih[19] & JBG_DPLAST ? " DPLAST" : "",
	bih[19] & 0x80 ? " other" : "");
    printf("		%u stripes, %d layers, %d planes\n",
	((yd >> bih[1]) +  ((((1UL << bih[1]) - 1) & xd) != 0) + l0 - 1) / l0,
	bih[1] - bih[0], bih[2]);
}

void
proff(int curOff)
{
    if (PrintOffset)
        printf("%d:     ", curOff);
    else if (PrintHexOffset)
        printf("%6x:    ", curOff);
}

void
decode(FILE *fp)
{
    int		c;
    int		rc;
    FILE	*dfp = NULL;
    int		pageNum = 0;
    int		i;
    int		curOff = 0;
    struct jbg_dec_state	s[5];
    unsigned char	bih[5][20];
    int			imageCnt[5] = {0,0,0,0};
    int         	pn = 0;
    int			ver, end;
    char		buf[5120*1024];

    c = getc(fp);
    if (c == EOF)
    {
	printf("EOF on file\n");
	return;
    }
    ungetc(c, fp);
    if (c == '\033')
    {
	while (fgets(buf, sizeof(buf), fp))
	{
	    proff(curOff);
	    if (buf[0] == '\033')
	    {
		printf("\\033");
		fputs(buf+1, stdout);
	    }
	    else
		fputs(buf, stdout);
	    curOff += strlen(buf);
	    if ((strcmp(buf, "@PJL ENTER LANGUAGE = QPDL\r\n") == 0)
		|| (strcmp(buf, "@PJL ENTER LANGUAGE = QPDL\n") == 0))
		break;
	}
    }

    for (;;)
    {
	int	reclen;
	int	rectype, subtype;
	int	wb, h, comp, stripe;
	int	cksum;

	#define STRARY(X, A) \
	    ((X) >= 0 && (X) < sizeof(A)/sizeof(A[0])) \
	    ? A[X] : "UNK"
	char *strsize[] = {
	    /*00*/ "letter", "legal", "a4", "exec", "ledger",
	    /*05*/ "a3", "env#10", "envMonarch", "envC5", "envDL",
	    /*10*/ "b4", "b5jis", "b5iso", "UNK", "UNK",
	    /*15*/ "UNK", "a5", "a6", "UNK", "UNK",
	    /*20*/ "UNK", "custom",  "UNK", "envC6", "folio",
	    /*25*/ "env6.75", "env#9",  "UNK", "oficio", "UNK",
	    /*30*/ "statement", "UNK",  "UNK", "UNK", "UNK",
	    };
	char *strsource[] = {
	    /*00*/ "unk", "auto", "manual", "multi", "tray1",
	    };

	rectype = fgetc(fp);
	if (rectype == EOF)
	    break;

	proff(curOff);
	curOff++;
	printf("RECTYPE 0x%x", rectype);

	switch (rectype)
	{
	case 0x01:
	    printf("	len=3\n");
	    if (fread(buf+1, 2, 1, fp) != 1)
		error(1, "Couldn't get 2 bytes\n");
	    curOff += 2;
	    printf("\t\tcopies=%d\n", getBEword(buf+1));
	    break;
	case 0x11:	// NOT JBIG!!
	    if (fread(buf+1, 4, 1, fp) != 1)
	    {
		printf("\n");
                error(1, "Couldn't get 4 bytes\n");
	    }
	    curOff += 4;
	    reclen = getBEdword(buf+1);
	    ++reclen;
	    printf("	len=%d(0x%x)\n", 5+reclen, 5+reclen);
	    if (fread(buf, reclen, 1, fp) != 1)
		error(1, "Couldn't get 0x%x(%d) bytes\n", reclen, reclen);
	    curOff += reclen;

	    break;
	case 0x09:
	    printf("	len=0\n");
	    goto done;
	    break;
	case 0x00:
	    printf("	len=17	pageNum=%d\n", ++pageNum);
	    if (fread(buf+1, 16, 1, fp) != 1)
		error(1, "Couldn't get 16 bytes\n");
	    curOff += 16;
	    printf("\t\tyres=%d, copies=%d, papersize=%s(%d), w=%d, h=%d\n",
		buf[1]*100,
		getBEword(buf+2),
		STRARY((int)buf[4], strsize), buf[4],
		getBEword(buf+5),
		getBEword(buf+7)
		);
	    printf("\t\tpapersource=%s, unk=%d, duplex=%d:%d, unk=%d,%d, "
		" unk=%d(0x%x)\n"
		, STRARY((int)buf[9], strsource)
		, buf[0xA]
		, buf[0xB]
		, buf[0xC]
		, buf[0xD]
		, buf[0xE]
		, getBEword(buf+15)
		, getBEword(buf+15)
		);
	    printf("\t\txres=%d\n"
		, (getBEword(buf+15) & 0xff) * 100
		);

	    pn = 0;
	    memset(imageCnt, 0, sizeof(imageCnt));
	    break;
	case 0x13:
	    printf("    len=15\n");
            if (fread(buf+1, 14, 1, fp) != 1)
                error(1, "Couldn't get 14 bytes\n");
            curOff += 14;
	    printf("\t\t");
	    for (i = 1; i <= 14; ++i)
		printf("%02x, ", buf[i]);
	    printf("\n");
	    break;
	case 0x14:
            if (fread(buf+1, 7, 1, fp) != 1)
                error(1, "Couldn't get 7 bytes\n");
	    curOff += 7;
	    subtype = buf[1];
	    if (subtype == 0x10)
	    {
		printf("    len=8\n");
		printf("\t\tunknown: ");
		for (i = 1; i <= 7; ++i)
		    printf("%02x, ", (unsigned char) buf[i]);
		printf("\n");
	    }
	    else
	    {
		/* BIH */
		printf("    len=25\n");
		if (fread(buf+7+1, 24-7, 1, fp) != 1)
		    error(1, "Couldn't get 24 bytes\n");
		curOff += 24-7;
		if (0)
		{
		    printf("\t\t");
		    for (i = 1; i <= 16; ++i)
			printf("%02x, ", (unsigned char) buf[i]);
		    printf("\n\t\t");
		    for (i = 17; i <= 24; ++i)
			printf("%02x, ", (unsigned char) buf[i]);
		}
		else
		{
		    printf("\t\t");
		    for (i = 21; i <= 24; ++i)
			printf("%02x, ", (unsigned char) buf[i]);
		    printf("(Margin=%d)", (unsigned char) buf[24]);
		}
		printf("\n");
		print_bih( (unsigned char *) buf+1);
		for (i = 0; i <=4; ++i)
		    memcpy(bih[i], buf+1, 20);
	    }
	    break;
	case 0x0c:
	    if (fread(buf+1, 11, 1, fp) != 1)
	    {
		printf("\n");
		error(1, "Couldn't get 11 bytes\n");
	    }
	    curOff += 11;
	    stripe = buf[1];
	    wb = getBEword(buf+2);
	    h = getBEword(buf+4);
	    pn = buf[6];
	    comp = buf[7];
	    reclen = getBEdword(buf+8);
	    if (comp == 0x12)
		reclen++;
	    if (comp == 0x11)
		reclen++;
	    printf("\tlen=%d(0x%x)\n", 12+reclen, 12+reclen);
	    printf("\t\tstripe=%d, WB=%d(0x%x), H=%d(0x%x), plane=%d, "
			"comp=0x%x,\n\t\tlen=%d(0x%x)\n",
			stripe, wb, wb, h, h, pn, comp, reclen, reclen);

	    if (fread(buf, reclen, 1, fp) != 1)
		error(1, "Couldn't get 0x%x(%d) bytes\n", reclen, reclen);
	    curOff += reclen;
	
	    cksum = 0;
	    for (i = 0; i < reclen-4; ++i)
		cksum += (unsigned char) buf[i];

	    if (comp == 0x15)
	    {
		if (0) printf("pn=%d\n", pn);
	    }
	    else if ( ((unsigned char) buf[0]) == 0xef)
	    {
		ver = getLEdword(buf+0) >> 28;
		end = (getBEdword(buf+8) >> 24) - 1;
		printf("\t\tmagic=0x%x, len=%d(0x%x), unk=%x,%x,%x,%x,%x,%x,"
		    "\n\t\tend=%d, ver=%d, checksum WANT=0x%x GOT=0x%x\n"
		    , getLEdword(buf+0)
		    , getLEdword(buf+4)
		    , getLEdword(buf+4)
		    , getLEdword(buf+8)
		    , getLEdword(buf+12)
		    , getLEdword(buf+16)
		    , getLEdword(buf+20)
		    , getLEdword(buf+24)
		    , getLEdword(buf+28)
		    , end
		    , ver
		    , cksum
		    , getLEdword(buf+reclen-4)
		    );
	    }
	    else
	    {
		ver = getBEdword(buf+0) >> 28;
		end = (getBEdword(buf+8) >> 24) - 1;
		printf("\t\tmagic=0x%x, len=%d(0x%x), unk=%x,%x,%x,%x,%x,%x,"
		    "\n\t\tend=%d, ver=%d, checksum WANT=0x%x GOT=0x%x\n"
		    , getBEdword(buf+0)
		    , getBEdword(buf+4)
		    , getBEdword(buf+4)
		    , getBEdword(buf+8)
		    , getBEdword(buf+12)
		    , getBEdword(buf+16)
		    , getBEdword(buf+20)
		    , getBEdword(buf+24)
		    , getBEdword(buf+28)
		    , end
		    , ver
		    , cksum
		    , getBEdword(buf+reclen-4)
		    );
	    }

	    if (comp == 0x13 || comp == 0x15)
	    {
if (0) printf("stripe=%d\n", stripe);
		// if ( (comp == 0x13 && stripe == 0) || comp == 0x15)
		if (comp == 0x13 && stripe == 0)
		{
		    size_t      cnt;

		    memcpy(bih[pn], buf+32, 20);
		    print_bih(bih[pn]);

		    jbg_dec_init(&s[pn]);
		    rc = jbg_dec_in(&s[pn], bih[pn], 20, &cnt);
		    if (rc == JBG_EIMPL)
			error(1, "JBIG uses unimpl feature\n");
		    break;
		}
		else if (comp == 0x15)
		{
		    size_t      cnt;

		    jbg_dec_init(&s[pn]);
		    rc = jbg_dec_in(&s[pn], bih[pn], 20, &cnt);
		    if (rc == JBG_EIMPL)
			error(1, "JBIG uses unimpl feature\n");
		}
		else if (comp == 0x13 && stripe >= 1)
		{
		    printf("\t\tData: ");
		    for (i = 0; i < 16; ++i)
			printf("%02x ", (unsigned char) buf[32+i]);
		    printf("...\n");
		}

		if (DecFile)
		{
		    if (comp == 0x15)
		    {
			reclen -= 4;
			i = 0;
		    }
		    else
			i = 32;
if (0 && comp == 0x15) printf("c=%02x ", (unsigned char) buf[i]);
		    for (; i < reclen; ++i)
		    {
			size_t              cnt;
			unsigned char       ch = c;

			ch = c = buf[i];

			rc = JBG_EAGAIN;
if (0 && comp == 0x15) printf("i=%d ", i);
			rc = jbg_dec_in(&s[pn], &ch, 1, &cnt);
if (0 && comp == 0x15) printf("rc=%d ", rc);
if (0 && comp == 0x15 && i == (reclen-1))  printf("c=%02x ", ch);
if (0 && comp == 0x15 && i == (reclen-2))  printf("c=%02x ", ch);
			if (rc == JBG_EOK)
			{
			    int     h, w, len;
			    unsigned char *image;

if (0 && comp == 0x15) printf("OK\n");
if (0) printf("OK\n");
			    // debug(0, "JBG_EOK: %d\n", pn);
			    h = jbg_dec_getheight(&s[pn]);
			    w = jbg_dec_getwidth(&s[pn]);
			    image = jbg_dec_getimage(&s[pn], 0);
			    len = jbg_dec_getsize(&s[pn]);
			    if (comp == 0x13 && image)
			    {
				char        buf[512];
				sprintf(buf, "%s-%02d-%d.pbm",
					DecFile, pageNum, pn);
				dfp = fopen(buf,
					    imageCnt[pn] ? "a" : "w");
				if (dfp)
				{
				    if (imageCnt[pn] == 0)
					fprintf(dfp, "P4\n%8d %8d\n", w, h);
				    //imageCnt[pn] += incrY;
				    rc = fwrite(image, 1, len, dfp);
				    fclose(dfp);
				    dfp = NULL;
				}
			    }
			    else if (comp == 0x15 && image)
			    {
				char        buf[512];
				if (dfp == 0)
				{
				    sprintf(buf, "%s-%02d-%d.pbm",
					DecFile, pageNum, pn);
				    dfp = fopen(buf,
					    imageCnt[pn] ? "r+" : "w");
				}
				if (dfp)
				{
				    fseek(dfp, 0, SEEK_SET);
				    // if (imageCnt[pn] == 0)
				    fprintf(dfp, "P4\n%8d %8d\n", w, h*stripe);
				    imageCnt[pn] += 1;
				    fseek(dfp, stripe * h * wb, SEEK_CUR);
				    rc = fwrite(image, 1, len, dfp);
				    fclose(dfp);
				    dfp = NULL;
				}
			    }
			    else
				debug(0, "Missing image %dx%d!\n", h, w);
			    jbg_dec_free(&s[pn]);
			    break;
			}
		    }
		}
	    }
	    break;
	default:
	    printf("\n");
	    error(1, "Unknown rectype 0x%x at 0x%x(%d)\n",
			rectype, curOff, curOff);
	    break;
	}
    }

done:
    c = fgetc(fp);
    if (c != 033)
	return;
    ungetc(c, fp);

    while (fgets(buf, sizeof(buf), fp))
    {
	proff(curOff);
	if (buf[0] == '\033')
        {
            printf("\\033");
            fputs(buf+1, stdout);
        }
        else
            fputs(buf, stdout);
	curOff += strlen(buf);
	if (strcmp(buf, "@PJL ENTER LANGUAGE=HIPERC\n") == 0)
	    break;
    }
    printf("\n");
}

int
main(int argc, char *argv[])
{
	extern int	optind;
	extern char	*optarg;
	int		c;

	while ( (c = getopt(argc, argv, "d:hor:D:?h")) != EOF)
		switch (c)
		{
		case 'd': DecFile = optarg; break;
		case 'r': RawFile = optarg; break;
		case 'o': PrintOffset = 1; break;
		case 'h': PrintHexOffset = 1; break;
		case 'D': Debug = atoi(optarg); break;
		default: usage(); exit(1);
		}

	argc -= optind;
	argv += optind;

	if (argc > 0)
	{
	    FILE	*fp;

	    fp = fopen(argv[0], "r");
	    if (!fp)
		error(1, "file '%s' doesn't exist\n", argv[0]);
	    decode(fp);
	    fclose(fp);
	}
	else
	    decode(stdin);

	exit(0);
}
